modules/adbookpspBidAdapter.js

```diff
@@ -0,0 +1,768 @@
+    .filter(validateBid(bidderRequestBody));
+  const targetingMap = buildTargetingMap(incomingBids);
+
+  return impBidsToPrebidBids(
+    incomingBids,
+    bidderRequestBody,
+    bidResponse.body.cur,
+    referrer,
+    targetingMap
+  );
+}
+
+function impBidsToPrebidBids(
+  incomingBids,
+  bidderRequestBody,
+  bidResponseCurrency,
+  referrer,
+  targetingMap
+) {
+  return incomingBids
+    .map(
+      impToPrebidBid(
+        bidderRequestBody,
+        bidResponseCurrency,
+        referrer,
+        targetingMap
+      )
+    )
+    .filter((i) => i !== null);
+}
+
+const impToPrebidBid =
+  (bidderRequestBody, bidResponseCurrency, referrer, targetingMap) => (bid) => {
+    try {
+      const bidRequest = findBidRequest(bidderRequestBody, bid);
+
+      if (!bidRequest) {
+        utils.logError(`${BIDDER_CODE}: Could not match bid to bid request`);
+
+        return null;
+      }
+      const categories = utils.deepAccess(bid, 'cat', []);
+      const mediaType = getMediaType(bid.adm);
+      let prebidBid = {
+        ad: bid.adm,
+        adId: bid.adid,
+        adserverTargeting: targetingMap[bid.impid],
+        adUnitCode: bidRequest.tagid,
+        bidderRequestId: bidderRequestBody.id,
+        bidId: bid.id,
+        cpm: bid.price,
+        creativeId: bid.crid || bid.id,
+        currency: bidResponseCurrency || getBidderConfig('defaultCurrency'),
+        height: bid.h,
+        lineItemId: utils.deepAccess(bid, 'ext.liid'),
+        mediaType,
+        meta: {
+          advertiserDomains: bid.adomain,
+          mediaType: mediaType,
+          primaryCatId: categories[0],
+          secondaryCatIds: categories.slice(1),
+        },
+        netRevenue: true,
+        nurl: bid.nurl,
+        referrer: referrer,
+        requestId: bid.impid,
+        ttl: getBidderConfig('bidTTL'),
+        width: bid.w,
+      };
+
+      if (mediaType === VIDEO) {
+        prebidBid = {
+          ...prebidBid,
+          ...getVideoSpecificParams(bidRequest, bid),
+        };
+      }
+
+      return prebidBid;
+    } catch (error) {
+      utils.logError(error);
+
+      return null;
+    }
+  };
+
+function getVideoSpecificParams(bidRequest, bid) {
+  return {
+    height: bid.h || bidRequest.video.h,
+    vastXml: bid.adm,
+    width: bid.w || bidRequest.video.w,
+  };
+}
+
+function buildTargetingMap(bids) {
+  const impIds = bids.map(({ impid }) => impid).filter(utils.uniques);
+  const values = impIds.reduce((result, id) => {
+    result[id] = {
+      lineItemIds: [],
+      dealIds: [],
+      adIds: [],
+    };
+
+    return result;
+  }, {});
+
+  bids.forEach((bid) => {
+    values[bid.impid].lineItemIds.push(bid.ext.liid);
+    values[bid.impid].dealIds.push(bid.dealid);
+    values[bid.impid].adIds.push(bid.adid);
+  });
+
+  const targetingMap = {};
+
+  for (const id of impIds) {
+    targetingMap[id] = {
+      hb_liid_adbookpsp: values[id].lineItemIds.join(TARGETING_VALUE_SEPARATOR),
```

Comment:
what is happening here

Reply:
Do you want to check said script into integration examples?

Reply:
I meant the script rendered by GAM, I'm not sure we're interested in adding an example right now. Is this something you require?

Reply:
No, but please add prominent disclosure in your doc repo pull and your .md file that custom gam creatives are required and that this bidder only works in its own order and that the prebid universal creative does not work etc. Please be very explicit and redundant.

Reply:
I've expanded the note about the adapter requirements in the doc pull request and brought the note to the readme file here.